x-pack/solutions/security/plugins/elastic_assistant/server/lib/defend_insights/evaluation/helpers/get_custom_evaluator/customPolicyResponseFailureEvaluator.ts

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { Example, Run } from 'langsmith';
+import type { EvaluatorT } from 'langsmith/evaluation';
+import type { DefendInsights, DefendInsight } from '@kbn/elastic-assistant-common';
+
+import { EVALUATOR_ERRORS } from './constants';
+
+export interface ExampleOutput {
+  insights: DefendInsights;
+}
+
+export function isValidExampleOutput(output: ExampleOutput): output is ExampleOutput {
+  // Check if output is an object and has the expected structure. It's defined in LangSmith, hence needs validation.
+  return (
+    output &&
+    Array.isArray(output.insights) &&
+    output.insights.every(
+      (insight) =>
+        typeof insight === 'object' &&
+        insight !== null &&
+        'group' in insight &&
+        'events' in insight &&
+        Array.isArray(insight.events) &&
+        'remediation' in insight &&
+        typeof insight.remediation?.message === 'string'
+    )
+  );
+}
+
+function failWithComment(comment: string): { key: string; score: number; comment: string } {
+  return {
+    key: 'correct',
+    score: 0,
+    comment,
+  };
+}
+
+function tokenize(text: string): string[] {
+  return text
+    .toLowerCase()
+    .replace(/[^\w\s]/g, ' ')
+    .split(/\s+/)
+    .filter((token) => token.length > 2);
+}
+
+function cosineSimilarity(text1: string, text2: string): number {
+  const tokens1 = tokenize(text1);
+  const tokens2 = tokenize(text2);
+
+  if (tokens1.length === 0 || tokens2.length === 0) {
+    return 0;
+  }
+
+  const allTokens = [...new Set([...tokens1, ...tokens2])];
+  const vector1 = allTokens.map((token) => tokens1.filter((t) => t === token).length);
+  const vector2 = allTokens.map((token) => tokens2.filter((t) => t === token).length);
+
+  const dotProduct = vector1.reduce((sum, a, i) => sum + a * vector2[i], 0);
+  const magnitude1 = Math.sqrt(vector1.reduce((sum, a) => sum + a * a, 0));
+  const magnitude2 = Math.sqrt(vector2.reduce((sum, a) => sum + a * a, 0));
+
+  return magnitude1 && magnitude2 ? dotProduct / (magnitude1 * magnitude2) : 0;
+}
```

Comment:
Well implemented 👍 . Maybe think about extracting this func to a more common place, since it's arguments are not tied to PRF.